File: solutions/prolog/space-age/1/space_age.pl
% Shree DR.MDD ko samarpit

space_age(Body, Seconds, Years) :-
    revolution_time(Body, Ratio),
    Years is Seconds / 31557600 / Ratio.

revolution_time("Mercury", 0.2408467).
revolution_time("Venus",   0.61519726).
revolution_time("Earth",   1.0).
revolution_time("Mars",    1.8808158).
revolution_time("Jupiter", 11.862615).
revolution_time("Saturn",  29.447498).
revolution_time("Uranus",  84.016846).
revolution_time("Neptune", 164.79132).

File: solutions/erlang/space-age/1/src/space_age.erl
% For my Junko F. Didi and Shree DR.MDD

-module(space_age).
-export([age/2]).

-define(EARTH_YEAR_SEC, 31557600).
-define(MERCURY_YEAR_SEC, ?EARTH_YEAR_SEC * 0.2408467).
-define(VENUS_YEAR_SEC, ?EARTH_YEAR_SEC * 0.61519726).
-define(MARS_YEAR_SEC, ?EARTH_YEAR_SEC * 1.8808158).
-define(JUPITER_YEAR_SEC, ?EARTH_YEAR_SEC * 11.862615).
-define(SATURN_YEAR_SEC, ?EARTH_YEAR_SEC * 29.447498).
-define(URANUS_YEAR_SEC, ?EARTH_YEAR_SEC * 84.016846).
-define(NEPTUNE_YEAR_SEC, ?EARTH_YEAR_SEC * 164.79132).

age(Planet, Seconds) ->
    Period = orbital_time(Planet),
    Seconds / Period.

orbital_time(Planet) ->
    proplists:get_value(Planet, all_orbits()).

all_orbits() ->
    [{earth, ?EARTH_YEAR_SEC},
     {mercury, ?MERCURY_YEAR_SEC},
     {venus, ?VENUS_YEAR_SEC},
     {mars, ?MARS_YEAR_SEC},
     {jupiter, ?JUPITER_YEAR_SEC},
     {saturn, ?SATURN_YEAR_SEC},
     {uranus, ?URANUS_YEAR_SEC},
     {neptune, ?NEPTUNE_YEAR_SEC}].
